spark/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -396,48 +397,64 @@ public void close() throws IOException {
+      Set<Integer> idColumns = Sets.newHashSet();
+      for (Integer i : spec.identitySourceIds()) {
+        if (spec.schema().columns().stream()
+            .noneMatch(j -> j.type().isStructType() && j.type().asStructType().field(i) != null)
+        ) {
+          idColumns.add(i);
+        }
+      }
 
-      // schema needed for the projection and filtering
-      StructType sparkType = SparkSchemaUtil.convert(finalSchema);
-      Schema requiredSchema = SparkSchemaUtil.prune(tableSchema, sparkType, task.residual(), caseSensitive);
+      Schema requiredSchema = SparkSchemaUtil.prune(tableSchema,
+          SparkSchemaUtil.convert(finalSchema), task.residual(), caseSensitive);
       boolean hasJoinedPartitionColumns = !idColumns.isEmpty();
       boolean hasExtraFilterColumns = requiredSchema.columns().size() != finalSchema.columns().size();
 
       Schema iterSchema;
       Iterator<InternalRow> iter;
 
       if (hasJoinedPartitionColumns) {
-        // schema used to read data files
         Schema readSchema = TypeUtil.selectNot(requiredSchema, idColumns);
         Schema partitionSchema = TypeUtil.select(requiredSchema, idColumns);
-        PartitionRowConverter convertToRow = new PartitionRowConverter(partitionSchema, spec);
-        JoinedRow joined = new JoinedRow();
 
-        InternalRow partition = convertToRow.apply(file.partition());
-        joined.withRight(partition);
+        Map<Integer, Object> partitionValueMap = Maps.newHashMap();
+        Map<String, Integer> partitionSpecFieldIndexMap = Maps.newHashMap();
+        Map<String, Class<?>> partitionSpecJavaTypeMap = Maps.newHashMap();
+        Map<String, DataType> partitionSpecDataTypeMap = Maps.newHashMap();
+        for (int i = 0; i < spec.fields().size(); i++) {
+          String fieldName = spec.fields().get(i).name();
+          partitionSpecFieldIndexMap.put(fieldName, i);
+          partitionSpecJavaTypeMap.put(fieldName, spec.javaClasses()[i]);
+          partitionSpecDataTypeMap.put(fieldName, getPartitionType(partitionSchema, fieldName));
+        }
 
-        // create joined rows and project from the joined schema to the final schema
-        iterSchema = TypeUtil.join(readSchema, partitionSchema);
-        iter = Iterators.transform(open(task, readSchema), joined::withLeft);
+        List<Types.NestedField> columns = new ArrayList<>(readSchema.columns());
+        for (Types.NestedField field : partitionSchema.columns()) {
+          int partitionIndex = partitionSpecFieldIndexMap.get(field.name());
+          DataType dataType = partitionSpecDataTypeMap.get(field.name());
+          Class<?> javaType = partitionSpecJavaTypeMap.get(field.name());
+          Object convertedValue = getPartitionValue(file.partition(), partitionIndex, dataType, javaType);
+          partitionValueMap.put(field.fieldId(), convertedValue);
+          columns.add(field);
+        }
 
+        columns.sort(Comparator.comparingInt(Types.NestedField::fieldId));
```

Comment:
Columns should not be sorted. What was the rationale for adding this?

Reply:
The partition columns are added at the end of  the schema thus leading to cast errors inside the projection dynamic code.

Reply:
I think that's probably caused by using the wrong schema to build the case. You should be building data with the required schema, then casting to the final schema.